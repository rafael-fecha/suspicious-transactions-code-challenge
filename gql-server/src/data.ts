export interface TransactionData {
  id: number;
  recipient: string;
  sender: string;
  amount: number;
  currency: string;
  locale: string;
  createdDateTimestamp: number;
  status: "Suspicious" | "Allowed" | "Blocked";
}

export const transactions: TransactionData[] = [
  {
    id: 1,
    recipient: "Rafa",
    sender: "John",
    amount: 100,
    currency: "USD",
    locale: "en-US",
    createdDateTimestamp: 51531515615156,
    status: "Suspicious"
  },
  {
    id: 2,
    recipient: "Rafa",
    sender: "John",
    amount: 100,
    currency: "EUR",
    locale: "de-DE",
    createdDateTimestamp: 51531515615156,
    status: "Suspicious"
  }
];
