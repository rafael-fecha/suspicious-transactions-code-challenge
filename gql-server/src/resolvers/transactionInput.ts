import { Field, InputType } from "type-graphql";

import { TransactionStatus } from "../interfaces/transaction.interface";

@InputType()
export class TransactionInput {
  @Field()
  id: number;

  @Field()
  status: TransactionStatus;
}
