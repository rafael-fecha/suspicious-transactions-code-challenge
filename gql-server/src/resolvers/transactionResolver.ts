import { Arg, Query, Mutation, Resolver } from "type-graphql";
import { TransactionData, transactions } from "../data";
import Transaction from "../schemas/Transaction";
import { TransactionInput } from "./transactionInput";

@Resolver(of => Transaction)
export class TransactionResolver {
  @Query(returns => [Transaction], { nullable: true })
  transactionSuspicious(): TransactionData[] | undefined {
    return transactions.filter(
      transaction => transaction.status === "Suspicious"
    );
  }

  @Mutation(returns => Transaction)
  markTransactionSuspicious(@Arg("transactionData")
  {
    id,
    status
  }: TransactionInput): TransactionData {
    const foundTransaction = transactions.find(t => t.id === id);
    if (!foundTransaction) {
      throw new Error(`Couldn't find transaction with id ${id}`);
    }

    foundTransaction.status = status;
    return foundTransaction;
  }
}
