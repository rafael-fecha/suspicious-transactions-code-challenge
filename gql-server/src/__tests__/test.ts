import "reflect-metadata";
import { buildSchema } from "type-graphql";

const { ApolloServer, gql } = require("apollo-server-express");
const { createTestClient } = require("apollo-server-testing");

import { TransactionResolver } from "../resolvers/transactionResolver";
import { transactions } from "../data";

it("should get only the suspicious transactions", async () => {
  const schema = await buildSchema({
    resolvers: [TransactionResolver],
    emitSchemaFile: true
  });

  const server = new ApolloServer({
    schema
  });

  const { query, mutate } = createTestClient(server);

  const {
    data: { transactionSuspicious }
  } = await query({
    query: gql`
      query {
        transactionSuspicious {
          id
          sender
          recipient
          amount
          currency
          locale
          createdDateTimestamp
          status
        }
      }
    `
  });

  const result = JSON.parse(JSON.stringify(transactionSuspicious));
  expect(result).toEqual(transactions);
});

it("should mark the transaction as blocked", async () => {
  const schema = await buildSchema({
    resolvers: [TransactionResolver],
    emitSchemaFile: true
  });

  const server = new ApolloServer({
    schema
  });

  const { mutate } = createTestClient(server);

  const {
    data: { markTransactionSuspicious }
  } = await mutate({
    mutation: gql`
      mutation MarkTransactionSuspicious($transactionData: TransactionInput!) {
        markTransactionSuspicious(transactionData: $transactionData) {
          id
          sender
          status
        }
      }
    `,
    variables: { transactionData: { id: 1, status: "Blocked" } }
  });

  const result = JSON.parse(JSON.stringify(markTransactionSuspicious));
  expect(result.status).toEqual("Blocked");
});
