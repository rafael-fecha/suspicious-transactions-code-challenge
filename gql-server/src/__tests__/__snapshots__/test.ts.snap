// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches single launch 1`] = `
Object {
  "data": Object {
    "transactionSuspicious": Array [
      Object {
        "amount": 100,
        "currency": "USD",
        "id": 1,
        "locale": "en-US",
        "recipient": "Rafa",
        "sender": "John",
      },
      Object {
        "amount": 100,
        "currency": "EUR",
        "id": 2,
        "locale": "de-DE",
        "recipient": "Rafa",
        "sender": "John",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
